rfd/0043-kubeaccess-multiparty.md

```diff
@@ -0,0 +1,386 @@
+
+    // Reason is an arbitrary string that may be used to describe the session and/or it's purpose.
+    string Reason = 7 [ (gogoproto.jsontag) = "reason,omitempty" ];
+
+    // Invited is a list of invited users, this field is interpreted by different
+    // clients on a best-effort basis and used for delivering notifications to invited users.
+    repeated string Invited = 8
+        [ (gogoproto.jsontag) = "invited,omitempty" ];
+}
+
+// SessionType encodes different types of sessions.
+enum SessionType {
+    // NONE is a placeholder variant and isn't valid.
+    NONE = 0;
+
+    // KUBERNETES means a session initiated via Kubernetes Access.
+    KUBERNETES = 1;
+
+    // SSH means a standard SSH session initiated via `tsh` or web.
+    SSH = 2;
+}
+
+// SessionState represents the state of a session.
+enum SessionState {
+    // PENDING variant represents a session that is waiting on participants to fulfill the criteria
+    // to start the session.
+    PENDING = 0;
+
+    // RUNNING variant represents a session that has had it's criteria for starting
+    // fulfilled at least once and has transitioned to a RUNNING state.
+    RUNNING = 1;
+
+    // TERMINATED variant represents a session that is no longer running and due for removal.
+    TERMINATED = 2;
+}
+```
+
+### Configurable Model Proposition
+
+Instead of having fixed fields for specifying values such as required session viewers and roles this
+model centers around conditional allow rules and filters. It is implemented as a bi-directional mapping between the role of the session initiator and the roles of the other session participants.
+
+Roles can have `require_session_join` rule under `allow` containing requirements for session participants
+before a session may be started with privilege access to nodes that the role provides.
+
+Roles can also have an `join_sessions` rule under `allow` specifying which roles
+and session types that that the role grants privileges to join.
+
+We will only initially support the modes `moderator` for Kubernetes Access and `peer` for SSH sessions.
+An `observer` mode also exists which only grants access to view but not terminate an ongoing session.
+
+Imagine you have 4 roles:
+- `prod-access`
+- `senior-dev`
+- `customer-db-maintenance`
+- `maintenance-observer`
+
+And these requirements:
+- `prod-access` should be able to start sessions of any type with either one `senior-dev` observeror two `dev` observers.
+- `senior-dev` should be able to start sessions of any type without oversight.
+- `customer-db-maintenance` needs oversight by one `maintenance-observer` on `ssh` type sessions.
+
+Then the 4 roles could be defined as follows:
+
+```yaml
+kind: role
+metadata:
+  name: prod-access
+spec:
+  allow:
+    require_session_join:
+      - name: Senior dev oversight
+        filter: 'contains(observer.roles,"senior-dev")'
+        kinds: ['k8s', 'ssh']
+        modes: ['moderator']
+        count: 1
+        require_mfa_tap: true
```

Comment:
we removed `require_mfa_tap` in favor of session option `per_session_mfa` right?

Reply:
Yep, I copy pasted them somewhere while editing and missed to remove them again.